{
 "cells": [
  {
   "cell_type": "raw",
   "id": "6ceef62f-d385-45c0-b9ca-67527a82dddf",
   "metadata": {},
   "source": [
    "Analysis of eTAC scRNAseq data\n",
    "author: Joe Germino (joe.germino@ucsf.edu)\n",
    "Created on Mar 31 2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3158aae4-e0e9-494c-8bad-6a071fbf1fc3",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7855fa83-efde-4a7f-8843-ca541791fca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import scanpy as sc\n",
    "import scrnatools as rna # custom helper methods for scRNAseq analysis available at https://github.com/j-germino/sc-rna-tools-git.git\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import numpy as np\n",
    "import scvi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a827f6-77b6-4723-8818-0daf477897e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "logger = rna.configs.create_logger(__name__)\n",
    "rna.configs.verbosity = \"warning\"\n",
    "sns.set_theme(context=\"paper\", style=\"white\", )\n",
    "plt.rcParams[\"figure.facecolor\"] = \"white\"\n",
    "plt.rcParams[\"axes.facecolor\"] = \"white\"\n",
    "plt.rcParams[\"savefig.facecolor\"] = \"white\"\n",
    "# Set working directory to project root folder\n",
    "if os.getcwd() != \"I:\\Desktop\\Joe\\Projects\\eTAC-analysis\":\n",
    "    os.chdir(\"I:\\Desktop\\Joe\\Projects\\eTAC-analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5fc4c07-150a-4e91-8681-3c0fceab0410",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c624e824-9734-4d34-ba7e-dec79920bd10",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825fd788-16d7-45f1-9927-eae0c5b2e607",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cellranger h5 files in 'data' subfolder\n",
    "\n",
    "# New sort data\n",
    "wtPos = sc.read_10x_h5(\"data/uncombined_WT_KO_data/WT_Pos_fiveprime_rna_filtered_feature_bc_matrix.h5\")\n",
    "wtPos.var_names_make_unique()\n",
    "wtPos.obs[\"mouse\"] = \"WT\"\n",
    "wtPos.obs[\"sort\"] = \"GFPp\"\n",
    "wtPos.obs[\"sample\"] = \"new data\"\n",
    "wtPos.layers[\"raw_counts\"] = wtPos.X.copy()\n",
    "\n",
    "wtRef = sc.read_10x_h5(\"data/uncombined_WT_KO_data/WT_Ref_fiveprime_rna_filtered_feature_bc_matrix.h5\")\n",
    "wtRef.var_names_make_unique()\n",
    "wtRef.obs[\"mouse\"] = \"WT\"\n",
    "wtRef.obs[\"sort\"] = \"unsorted\"\n",
    "wtRef.obs[\"sample\"] = \"new data\"\n",
    "wtRef.layers[\"raw_counts\"] = wtRef.X.copy()\n",
    "\n",
    "# Published data (Wang et al, Science Immunol. 2021) DOI: 10.1126/sciimmunol.abl5053\n",
    "wtPosOld = sc.read_10x_h5(\"data/eTAC_old_data/RNA_GFP_filtered_feature_bc_matrix.h5\")\n",
    "wtPosOld.var_names_make_unique()\n",
    "wtPosOld.obs[\"mouse\"] = \"WT\"\n",
    "wtPosOld.obs[\"sort\"] = \"GFPp\"\n",
    "wtPosOld.obs[\"sample\"] = \"Wang, et al. 2021\"\n",
    "wtPosOld.layers[\"raw_counts\"] = wtPosOld.X.copy()\n",
    "\n",
    "wtRefOld = sc.read_10x_h5(\"data/eTAC_old_data/RNA_WT_filtered_feature_bc_matrix.h5\")\n",
    "wtRefOld.var_names_make_unique()\n",
    "wtRefOld.obs[\"mouse\"] = \"WT\"\n",
    "wtRefOld.obs[\"sort\"] = \"unsorted\"\n",
    "wtRefOld.obs[\"sample\"] = \"Wang, et al. 2021\"\n",
    "wtRefOld.layers[\"raw_counts\"] = wtRefOld.X.copy()\n",
    "\n",
    "eTACdict = {\"WT_pos\": wtPos, \"WT_ref\": wtRef, \"WT_pos_old\": wtPosOld, \"WT_ref_old\": wtRefOld}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c40cd7fe-5d48-48d1-ab27-1d3a15b4152e",
   "metadata": {},
   "source": [
    "# Doublet filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b35f1e34-aa47-4a7a-9de6-cb6e6e354709",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter doublets using scrublet\n",
    "eTACdict = rna.qc.doublet_filter(\n",
    "    eTACdict,\n",
    "    score_threshold=0.3,\n",
    "    save_path=\"analysis/qc/scrublet/WT_LN_merged/\",\n",
    "    dpi=300,\n",
    "    context=\"paper\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e299732-d7e0-41d4-aae5-16e8062b2932",
   "metadata": {},
   "source": [
    "# Merge Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44738ed3-26cf-49cf-b17a-15a1271906d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "lnData = rna.tl.merge_anndata_dict(eTACdict, join=\"inner\")\n",
    "lnData.obs[\"batch_key\"] = lnData.obs.batch # Batch key is each individual sample (old v new + sort condition)\n",
    "lnData.write(\"processed_data/h5ad_files/WT_LN_merged_raw.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b42b245-6fb5-4918-9699-25c025e2a43d",
   "metadata": {},
   "source": [
    "# QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e831692-cfc6-4b18-986c-63206b90034d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_raw.h5ad\")\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\")\n",
    "sc.pp.calculate_qc_metrics(adata, percent_top=None, qc_vars=[\"mt\"], log1p=False, inplace=True)\n",
    "sc.pp.filter_cells(adata, min_counts=10)\n",
    "sc.pp.filter_genes(adata, min_counts=1)\n",
    "logger.info(f\"Num cells before filter: {len(adata.obs)}\")\n",
    "adata = rna.qc.filter_cells(\n",
    "    adata,\n",
    "    gene_thresholds=(600,5000),\n",
    "    count_thresholds=(1000,30000),\n",
    "    save_path=\"analysis/qc/WT_LN_merged\",\n",
    "    hue_key=\"batch\",\n",
    "    title=\"WT LN combined\",\n",
    "    dpi=300,\n",
    "    context=\"paper\",\n",
    "    dims=(18,5),\n",
    "    s=3,\n",
    ")\n",
    "logger.info(f\"Num cells after filter: {len(adata.obs)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07631395-65f0-4221-a358-b542bce48d73",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52686843-0e74-4db9-9115-35e7e4ae85c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "logger.info(\"Normalizing and log-transforming counts\")\n",
    "# Normalize counts per cell\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "# Log-transform counts\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "# Save raw, log transformed counts\n",
    "adata.raw = adata\n",
    "\n",
    "logger.info(\"Regress out ['total_counts', 'pct_counts_mt'] and scale\")\n",
    "# Regress and scale all genes\n",
    "sc.pp.regress_out(adata, [\"total_counts\", \"pct_counts_mt\"])\n",
    "sc.pp.scale(adata, max_value=10)\n",
    "\n",
    "# Save the preprocessed data for scVI training\n",
    "adata.write(\"processed_data/h5ad_files/WT_LN_merged_preprocessed.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22da93af-ade4-465a-b93e-132ac34a3c32",
   "metadata": {},
   "source": [
    "# scVI training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01416e4f-2101-4dc8-9c03-56bb6bf94af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure path for scVI models exists\n",
    "if not os.path.exists(\"processed_data/scVI_models/\"):\n",
    "        os.makedirs(\"processed_data/scVI_models/\")\n",
    "\n",
    "# Train model using raw counts, batch key as sample + sort combination (req. GPU)\n",
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_preprocessed.h5ad\")\n",
    "scvi.model.SCVI.setup_anndata(adata, layer=\"raw_counts\", batch_key=\"batch_key\")\n",
    "model = scvi.model.SCVI(adata)\n",
    "model.train(use_gpu=True)\n",
    "model.save(\"processed_data/scVI_models/WT_LN_merged\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6058db48-4507-4580-9113-9f1b477c3557",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98b3a274-83e3-4a95-8e0c-a2b07bccac8c",
   "metadata": {},
   "source": [
    "## Make UMAP, tSNE embeddings, get normalized expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af91518-5f38-4558-b0f5-12ff5af8aeb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Work from files saved for scVI training\n",
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_preprocessed.h5ad\")\n",
    "model = scvi.model.SCVI.load(\n",
    "    \"processed_data/scVI_models/WT_LN_merged\",\n",
    "    adata\n",
    ")\n",
    "# Get latent representation\n",
    "adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "# Save scVI normalized expression to a new layer, normalized to 10e4 counts per cell\n",
    "adata.layers[\"scVI_normalized\"] = model.get_normalized_expression(library_size=10e4)\n",
    "# Calculate neighbors using scVI latent space\n",
    "sc.pp.neighbors(adata, use_rep=\"X_scVI\")\n",
    "# Calculate UMAP, tSNE coordinates for each cell\n",
    "sc.tl.umap(adata, min_dist=0.5, spread=1)\n",
    "sc.tl.tsne(adata, use_rep=\"X_scVI\")\n",
    "# Save processed data as a new h5ad file\n",
    "adata.write(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54f264f7-60b8-4b99-8346-7bbe82efe96f",
   "metadata": {},
   "source": [
    "## Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1375b52-13fe-4a08-9c30-5dc75b1d52ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")\n",
    "sc.tl.leiden(adata, key_added=\"leiden\", resolution=0.8)\n",
    "adata.write(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "583546b2-1c64-46a5-90cf-a93fb63775dc",
   "metadata": {},
   "source": [
    "## Cell type annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659af818-ab2a-4532-a77f-850b4404fad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")\n",
    "adata.obs[\"cell_type\"] = adata.obs.leiden\n",
    "# Remove noise/junk cell clusters\n",
    "adata = adata[~adata.obs.cell_type.isin([\"23\", \"24\", \"29\"])].copy()\n",
    "# Relabel leiden cluster numbers as cell types\n",
    "cellTypes = {\n",
    "    \"18\": \"JC2\",\n",
    "    \"5\": \"JC1\",\n",
    "    \"15\": \"JC3\",\n",
    "    \"8\": \"mDC2-5\",\n",
    "    \"2\": \"mDC2-4\",\n",
    "    \"4\": \"mDC2-3\",\n",
    "    \"13\": \"mDC2-2\",\n",
    "    \"1\": \"mDC2-1\",\n",
    "    \"0\": \"mDC1-2\",\n",
    "    \"3\": \"mDC1-1\",\n",
    "    \"11\": \"mDC1-3\",\n",
    "    \"17\": \"ILC3\",\n",
    "    \"12\": \"CD4 T\",\n",
    "    \"14\": \"CD8 T\",\n",
    "    \"16\": \"B\",\n",
    "    \"20\": \"Fibro1\",\n",
    "    \"28\": \"Fibro3\",\n",
    "    \"27\": \"Fibro2\",\n",
    "    \"6\": \"MF\",\n",
    "    \"9\": \"cDC2\",\n",
    "    \"10\": \"cDC1\",\n",
    "    \"21\": \"Plasma cell\",\n",
    "    \"25\": \"pDC\",\n",
    "    \"19\": \"Mo\",\n",
    "    \"7\": \"BEC\",\n",
    "    \"26\": \"ILC\",\n",
    "    \"22\": \"LEC\",\n",
    "}\n",
    "adata.obs.cell_type = adata.obs.cell_type.replace(cellTypes)\n",
    "\n",
    "# Feature plots for cell type annotation\n",
    "rna.pl.sc_umap(\n",
    "    adata,\n",
    "    [\"cell_type\", \"sort\", \"Aire\", \"H2-Aa\", \"Ccr7\", \"Rorc\", \"Irf4\", \"Irf8\", \"Ptprc\", \"Cd8a\", \"Cd4\", \"Cd3e\", 'Cd19', \"Itgae\", \"Col1a1\", \"Loxl1\", \"Lum\", \"Fbln1\", \"Siglech\", \"Cd79a\", \n",
    "     \"Mki67\", \"Cd69\",\"Igha\", \"Xcr1\",\"Cd207\",\"Pld4\", \"Ccr9\", \"Sirpa\", \"Clec4n\", \"Il1b\",\"Cd34\", \"Adgre1\", \"Itgax\", \"Bst2\", \"Ly6c1\", \"Cd14\", \"Cd68\", \"Ccr5\", \"Il7r\", \"Sdc1\", \"Itgam\"],\n",
    "    legend_loc=[\"on_data\", \"bottom\"],\n",
    "    cmaps=[\"gist_rainbow\", \"tab10\", \"viridis\"],\n",
    "    gene_data=\"scVI_normalized\",\n",
    "    latent_rep=\"X_tsne\",\n",
    "    context=\"paper\",\n",
    "    dims=(6, 6),\n",
    "    s=5,\n",
    "    save_path=\"analysis/feature_plots/WT_LN_cell_type_features.png\",\n",
    "    dpi=300,\n",
    "    on_data_text_size=10,\n",
    "    legend_size=10,\n",
    ")\n",
    "adata.write(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09f6f37c-28c4-431c-a60b-a6ad7cf35634",
   "metadata": {},
   "source": [
    "## DE genes all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88450658-1acb-410d-98b9-28983a482e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"processed_data/h5ad_files/WT_LN_merged_scVI_processed.h5ad\")\n",
    "model = scvi.model.SCVI.load(\n",
    "    \"processed_data/scVI_models/WT_LN_merged\",\n",
    "    adata\n",
    ")\n",
    "# Get DE markers for cell type clusters using scVI differential expression method\n",
    "# default thresholds for marker genes: lfc_mean > 0, bayes_factor > 3, non_zeros_proportion > 0.1\n",
    "markers = rna.tl.cluster_de(\n",
    "    adata,\n",
    "    model,\n",
    "    cluster_key=\"cell_type\",\n",
    "    save_path=\"analysis/DE_genes/WT_LN_merged/cell_type_markers/\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcb859a3-d00b-4b7d-9119-25ae323c7bb6",
   "metadata": {},
   "source": [
    "## Embeddings, all clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bc601fd-a347-4270-aa16-9e37d415fec0",
   "metadata": {},
   "source": [
    "### Sample breakdown, all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6811fb86-350a-4b43-90c2-6d0cb73a4927",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.pl.sc_umap(\n",
    "    adata,\n",
    "    [\"sample\"],\n",
    "    cmaps=[\"tab10\"],\n",
    "    legend_loc=[ \"bottom\"],\n",
    "    latent_rep=\"X_tsne\",\n",
    "    ncols=1,\n",
    "    legend_size=14,\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=14,\n",
    "    title_size=14,\n",
    "    save_path=\"analysis/all_clusters_sample_source.pdf\", \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337e011c-9182-4e2b-851d-a115617d8072",
   "metadata": {},
   "source": [
    "### Sort breakdown, all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3695d1c-f179-4ca7-b600-9ae08b91ff32",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.pl.sc_umap(\n",
    "    adata,\n",
    "    [\"sort\"],\n",
    "    cmaps=[\"tab10\"],\n",
    "    legend_loc=[ \"bottom\"],\n",
    "    latent_rep=\"X_tsne\",\n",
    "    ncols=1,\n",
    "    legend_size=14,\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=14,\n",
    "    title_size=14,\n",
    "    save_path=\"analysis/all_clusters_sort.pdf\", \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414d8191-db1a-4cfb-8fff-0fa7ddfdd8eb",
   "metadata": {},
   "source": [
    "### Feature plots, all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b5cfe0-2df0-4265-af29-0ffbaee3837e",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression = rna.tl.get_expression_matrix(adata, \"scVI_normalized\")\n",
    "# Clip the bottom and top 1% of expression values for each gene to prevent outliers from affecting visualization\n",
    "thresholds = [\n",
    "    None,\n",
    "    (expression[\"Aire\"].quantile(0.01), expression[\"Aire\"].quantile(0.99)),\n",
    "    (expression[\"Rorc\"].quantile(0.01), expression[\"Rorc\"].quantile(0.99)),\n",
    "    (expression[\"Itgav\"].quantile(0.01), expression[\"Itgav\"].quantile(0.99)),\n",
    "    (expression[\"Itgb8\"].quantile(0.01), expression[\"Itgb8\"].quantile(0.99)),\n",
    "    (expression[\"Ccr7\"].quantile(0.01), expression[\"Ccr7\"].quantile(0.99)),\n",
    "    (expression[\"Cd80\"].quantile(0.01), expression[\"Cd80\"].quantile(0.99)),\n",
    "    (expression[\"Cd86\"].quantile(0.01), expression[\"Cd86\"].quantile(0.99)),\n",
    "    (expression[\"H2-Aa\"].quantile(0.01), expression[\"H2-Aa\"].quantile(0.99)),\n",
    "    (expression[\"H2-Ab1\"].quantile(0.01), expression[\"H2-Ab1\"].quantile(0.99)),\n",
    "    (expression[\"Klf4\"].quantile(0.01), expression[\"Klf4\"].quantile(0.99)),\n",
    "    (expression[\"Myc\"].quantile(0.01), expression[\"Myc\"].quantile(0.99)),\n",
    "    (expression[\"Ccr6\"].quantile(0.01), expression[\"Ccr6\"].quantile(0.99)),\n",
    "    (expression[\"Cd40\"].quantile(0.01), expression[\"Cd40\"].quantile(0.99)),\n",
    "    (expression[\"Cxcr6\"].quantile(0.01), expression[\"Cxcr6\"].quantile(0.99)),\n",
    "    (expression[\"Il7r\"].quantile(0.01), expression[\"Il7r\"].quantile(0.99)),\n",
    "    (expression[\"Thy1\"].quantile(0.01), expression[\"Thy1\"].quantile(0.99)),\n",
    "    (expression[\"Cd44\"].quantile(0.01), expression[\"Cd44\"].quantile(0.99)),\n",
    "    (expression[\"Il2ra\"].quantile(0.01), expression[\"Il2ra\"].quantile(0.99)),\n",
    "]\n",
    "rna.pl.sc_umap(\n",
    "    adata,\n",
    "    [\"cell_type\", \"Aire\", \"Rorc\", \"Itgav\", \"Itgb8\", \"Ccr7\", \"Cd80\", \"Cd86\", \"H2-Aa\", \"H2-Ab1\", \"Klf4\", \"Myc\", \"Ccr6\", \"Cd40\", \"Cxcr6\", \"Il7r\", \"Thy1\", \"Cd44\", \"Il2ra\"],\n",
    "    cmaps=[\"tab20\", \"tab10\", \"viridis\"],\n",
    "    legend_loc=[\"None\"],\n",
    "    latent_rep=\"X_tsne\",\n",
    "    gene_data=\"scVI_normalized\",\n",
    "    ncols=3,\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=21,\n",
    "    title_size=21,\n",
    "    save_path=\"analysis/WT_merged_feature_plots/littman_paper_all_clusters.png\",\n",
    "    cmap_label=\"Normalized expression\",\n",
    "    dpi=300,\n",
    "    thresholds=thresholds,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65e9c70a-2e78-45a8-8adb-eb69bdfa48cf",
   "metadata": {},
   "source": [
    "## Cosine similarity, all clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eb4406c-14ab-49dc-9946-b976b6f05152",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only need to run once to download immgen data, doesn't work on windows without installing wget\n",
    "!wget \"https://gist.github.com/vasilisNt/5e23eeefc188e1e772f428c74ef43277/raw/67f83d282b0b2180a8eeff74edf079d8826b12ba/immgen.tar.gz\"\n",
    "!tar -xzf immgen.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "770a6dfe-d7c6-448d-b6f7-010b26e51b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type similarity method uses the '.raw' attribute, make sure it contains raw counts (not log normalized\n",
    "adataCopy = adata.copy()\n",
    "del adataCopy.raw\n",
    "adataCopy.X = adataCopy.layers[\"raw_counts\"]\n",
    "adataCopy.raw = adataCopy\n",
    "\n",
    "# Add immgen similarity scores to adataCopy\n",
    "adataCopy = rna.tl.cell_type_similarity(\n",
    "    adataCopy,\n",
    "    \"immgen\",\n",
    "    save_path=\"analysis/all_WT_LN_similarity\",\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=21,\n",
    "    title_size=21,\n",
    "    cmap_label=\"Cosine similarity score\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8fcb4d2-e16f-4058-8a81-4e54c668c933",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataDict = {}\n",
    "# Immgen samples of interest\n",
    "keys = [\n",
    "    'DC_8-_Th_imm_score',\n",
    "    'DC_8+_Th_imm_score',\n",
    "    'DC_4+_Sp_imm_score',\n",
    "    'DC_8+_Sp_imm_score',\n",
    "    'DC_8-4-11b-_Sp_imm_score',\n",
    "    'DC_8-4-11b+_Sp_imm_score',\n",
    "    'DC_4+_SLN_imm_score',\n",
    "    'DC_8+_SLN_imm_score',\n",
    "    'DC_8-4-11b-_SLN_imm_score',\n",
    "    'DC_8-4-11b+_SLN_imm_score',\n",
    "    'DC_4+_MLN_imm_score',\n",
    "    'DC_8+_MLN_imm_score',\n",
    "    'DC_8-4-11b-_MLN_imm_score',\n",
    "    'DC_8-4-11b+_MLN_imm_score',\n",
    "    'DC_103-11b+_Lv_imm_score',\n",
    "    'DC_103+11b-_Lv_imm_score',\n",
    "    'DC_103+11b-_LuLN_imm_score',\n",
    "    'DC_103-11b+_LuLN_imm_score',\n",
    "    'DC_103-11b+24+_Lu_imm_score',\n",
    "    'DC_103+11b-_Lu_imm_score',\n",
    "    'DC_103-11b+F4_imm_score',\n",
    "    'DC_103+11b-_SI_imm_score',\n",
    "    'DC_103+11b+_SI_imm_score',\n",
    "    'DC_pDC_8-_Sp_imm_score',\n",
    "    'DC_pDC_8+_Sp_imm_score',\n",
    "    'DC_pDC_8+_SLN_imm_score',\n",
    "    'DC_pDC_8+_MLN_imm_score',\n",
    "    'DC_IIhilang-103-11blo_SLN_imm_score',\n",
    "    'DC_IIhilang-103-11b+_SLN_imm_score',\n",
    "    'DC_IIhilang+103+11blo_SLN_imm_score',\n",
    "    'DC_IIhilang+103-11b+_SLN_imm_score',\n",
    "    'DC_LC_Sk_imm_score',\n",
    "    'Ep_MEChi_Th_imm_score',\n",
    "    'Fi_MTS15+_Th_imm_score',\n",
    "    'Fi_Sk_imm_score',\n",
    "    'FRC_MLN_imm_score',\n",
    "    'FRC_SLN_imm_score',\n",
    "    'LEC_MLN_imm_score',\n",
    "    'LEC_SLN_imm_score',\n",
    "    'BEC_MLN_imm_score',\n",
    "    'BEC_SLN_imm_score',\n",
    "    'St_31-38-44-_SLN_imm_score',\n",
    "    'proB_CLP_BM_imm_score',\n",
    "    'proB_CLP_FL_imm_score',\n",
    "    'proB_FrA_BM_imm_score',\n",
    "    'proB_FrA_FL_imm_score',\n",
    "    'proB_FrBC_BM_imm_score',\n",
    "    'proB_FrBC_FL_imm_score',\n",
    "    'preB_FrC_BM_imm_score',\n",
    "    'preB_FrD_BM_imm_score',\n",
    "    'preB_FrD_FL_imm_score',\n",
    "    'B_FrE_BM_imm_score',\n",
    "    'B_FrE_FL_imm_score',\n",
    "    'B_T1_Sp_imm_score',\n",
    "    'B_T2_Sp_imm_score',\n",
    "    'B_T3_Sp_imm_score',\n",
    "    'B_Fo_Sp_imm_score',\n",
    "    'B_Fo_MLN_imm_score',\n",
    "    'B_Fo_LN_imm_score',\n",
    "    'B_Fo_PC_imm_score',\n",
    "    'B_GC_Sp_imm_score',\n",
    "    'B_MZ_Sp_imm_score',\n",
    "    'B_FrF_BM_imm_score',\n",
    "    'B1b_PC_imm_score',\n",
    "    'B1a_PC_imm_score',\n",
    "    'B1a_Sp_imm_score',\n",
    "    'GN_BM_imm_score',\n",
    "    'GN_Bl_imm_score',\n",
    "    'MF_BM_imm_score',\n",
    "    'MF_RP_Sp_imm_score',\n",
    "    'MF_Lu_imm_score',\n",
    "    'MF_103-11b+24-_Lu_imm_score',\n",
    "    'MF_II+480lo_PC_imm_score',\n",
    "    'MF_103-11b+_SI_imm_score',\n",
    "    'MF_11cloSer_SI_imm_score',\n",
    "    'MF_II-480hi_PC_imm_score',\n",
    "    'MF_Microglia_CNS_imm_score',\n",
    "    'Mo_6C+II-_BM_imm_score',\n",
    "    'Mo_6C-II-_BM_imm_score',\n",
    "    'Mo_6C+II-_Bl_imm_score',\n",
    "    'Mo_6C+II+_Bl_imm_score',\n",
    "    'Mo_6C-II-_Bl_imm_score',\n",
    "    'Mo_6C-II+_Bl_imm_score',\n",
    "    'Mo_6C-IIint_Bl_imm_score',\n",
    "    'Mo_6C+II-_LN_imm_score',\n",
    "    'NK_Sp_imm_score',\n",
    "    'NK_49CI-_Sp_imm_score',\n",
    "    'NK_49CI+_Sp_imm_score',\n",
    "    'NK_49H-_Sp_imm_score',\n",
    "    'NK_49H+_Sp_imm_score',\n",
    "    'NK_b2m-_Sp_imm_score',\n",
    "    'NK_DAP10-_Sp_imm_score',\n",
    "    'NK_DAP12-_Sp_imm_score',\n",
    "    'NK_CD127-_Sp_imm_score',\n",
    "    'NK_CD49b+_Lv_imm_score',\n",
    "    'NK_CD127-_SI_imm_score',\n",
    "    'ILC1_CD127+_Sp_imm_score',\n",
    "    'ILC1_CD49b-_Lv_imm_score',\n",
    "    'ILC1_CD127+_SI_imm_score',\n",
    "    'ILC2_SI_imm_score',\n",
    "    'ILC3_NKp46+_Rorgthi_SI_imm_score',\n",
    "    'ILC3_NKp46-_4-_SI_imm_score',\n",
    "    'ILC3_NKp46-_4+_SI_imm_score',\n",
    "    'preT_ETP_Th_imm_score',\n",
    "    'preT_ETP-2A_Th_imm_score',\n",
    "    'preT_DN2_Th_imm_score',\n",
    "    'preT_DN2A_Th_imm_score',\n",
    "    'preT_DN2B_Th_imm_score',\n",
    "    'preT_DN2-3_Th_imm_score',\n",
    "    'preT_DN3A_Th_imm_score',\n",
    "    'preT_DN3B_Th_imm_score',\n",
    "    'preT_DN3-4_Th_imm_score',\n",
    "    'T_DN4_Th_imm_score',\n",
    "    'T_ISP_Th_imm_score',\n",
    "    'T_DP_Th_imm_score',\n",
    "    'T_DPbl_Th_imm_score',\n",
    "    'T_DPsm_Th_imm_score',\n",
    "    'T_DP69+_Th_imm_score',\n",
    "    'T_4+8int_Th_imm_score',\n",
    "    'T_4SP69+_Th_imm_score',\n",
    "    'T_4SP24int_Th_imm_score',\n",
    "    'T_4SP24-_Th_imm_score',\n",
    "    'T_4int8+_Th_imm_score',\n",
    "    'T_8SP69+_Th_imm_score',\n",
    "    'T_8SP24int_Th_imm_score',\n",
    "    'T_8SP24-_Th_imm_score',\n",
    "    'T_4Nve_Sp_imm_score',\n",
    "    'T_4Mem_Sp_imm_score',\n",
    "    'T_4Mem44h62l_Sp_imm_score',\n",
    "    'T_4Nve_LN_imm_score',\n",
    "    'T_4Mem_LN_imm_score',\n",
    "    'T_4Mem44h62l_LN_imm_score',\n",
    "    'T_4Nve_PP_imm_score',\n",
    "    'T_4Nve_MLN_imm_score',\n",
    "    'T_4_LN_BDC_imm_score',\n",
    "    'T_4_PLN_BDC_imm_score',\n",
    "    'T_4_Pa_BDC_imm_score',\n",
    "    'T_4FP3-_Sp_imm_score',\n",
    "    'T_4FP3+25+_Sp_imm_score',\n",
    "    'T_4FP3+25+_AA_imm_score',\n",
    "    'T_4FP3+25+_LN_imm_score',\n",
    "    'T_8Nve_Sp_imm_score',\n",
    "    'T_8Mem_Sp_imm_score',\n",
    "    'T_8Nve_LN_imm_score',\n",
    "    'T_8Mem_LN_imm_score',\n",
    "    'T_8Nve_PP_imm_score',\n",
    "    'T_8Nve_MLN_imm_score',\n",
    "    'NKT_44-NK1_1-_Th_imm_score',\n",
    "    'NKT_44+NK1_1-_Th_imm_score',\n",
    "    'NKT_44+NK1_1+_Th_imm_score',\n",
    "    'NKT_4+_Sp_imm_score',\n",
    "    'NKT_4-_Sp_imm_score',\n",
    "    'NKT_4+_Lv_imm_score',\n",
    "    'NKT_4-_Lv_imm_score',\n",
    "    'Tgd_Th_imm_score',\n",
    "    'Tgd_vg1+vd6-24ahi_Th_imm_score',\n",
    "    'Tgd_vg1+vd6+24ahi_Th_imm_score',\n",
    "    'Tgd_vg2+24ahi_Th_imm_score',\n",
    "    'Tgd_vg2+24ahi_e17_Th_imm_score',\n",
    "    'Tgd_vg3+24ahi_e17_Th_imm_score',\n",
    "    'Tgd_vg5+24ahi_Th_imm_score',\n",
    "    'Tgd_vg1+vd6-24alo_Th_imm_score',\n",
    "    'Tgd_vg1+vd6+24alo_Th_imm_score',\n",
    "    'Tgd_vg2+24alo_Th_imm_score',\n",
    "    'Tgd_vg3+24alo_e17_Th_imm_score',\n",
    "    'Tgd_Sp_imm_score',\n",
    "    'Tgd_vg2-_Sp_imm_score',\n",
    "    'Tgd_vg2-_act_Sp_imm_score',\n",
    "    'Tgd_vg2+_Sp_imm_score',\n",
    "    'Tgd_vg2+_act_Sp_imm_score',\n",
    "    'Tgd_vg2-_Sp_TCRbko_imm_score',\n",
    "    'Tgd_vg2+_Sp_TCRbko_imm_score',\n",
    "    'Tgd_vg5-_IEL_imm_score',\n",
    "    'Tgd_vg5+_IEL_imm_score',\n",
    "    'Tgd_vg5-_act_IEL_imm_score',\n",
    "    'Tgd_vg5+_act_IEL_imm_score',\n",
    "    ]\n",
    "for key in keys:\n",
    "    dataDict[key.rsplit(\"_\", 2)[0]] = adataCopy.obs[key]\n",
    "dataDict[\"cell_type\"] = adataCopy.obs.cell_type\n",
    "data = pd.DataFrame(dataDict)\n",
    "cats = [\"JC1\", \"JC2\", \"JC3\", \"mDC1-1\", \"mDC1-2\", \"mDC1-3\", \"mDC2-1\", \"mDC2-2\", \"mDC2-3\", \"mDC2-4\", \"mDC2-5\", \"cDC1\", \"cDC2\", \"pDC\", \"MF\", \"Mo\", \"BEC\", \"LEC\", \"Fibro1\",\n",
    "        \"Fibro2\", \"Fibro3\", \"ILC\", \"ILC3\", \"CD4 T\", \"CD8 T\", \"B\", \"Plasma cell\"]\n",
    "data.cell_type.cat.reorder_categories(cats, inplace=True)\n",
    "# Average similarity scores bycell type\n",
    "data = data.groupby(\"cell_type\").mean()\n",
    "# Standardize scores to 0-1 for each immgen sample\n",
    "data = data.subtract(data.min(axis=0), axis=1)\n",
    "data = data.div(data.max(axis=0), axis=1)\n",
    "# Clustered heatmap of data\n",
    "ax = sns.clustermap(\n",
    "    data,\n",
    "    cmap=\"coolwarm\",\n",
    "    row_cluster=True,\n",
    "    col_cluster=True,\n",
    "    figsize=(65,16),\n",
    "    cbar_pos=None,\n",
    ")\n",
    "ax.ax_heatmap.tick_params(axis=\"both\", which=\"both\", labelsize=21)\n",
    "ax.ax_heatmap.set_ylabel(\"\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"analysis/all_cluster_similarity_heatmap_clustered.png\", dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c98f57ca-49f8-42f1-a1c6-6b8c39cd75c0",
   "metadata": {},
   "source": [
    "## Subclustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e407c9e6-b623-456d-87e6-f96d0b600359",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset on JCs, ILC3\n",
    "adataSubset = adata[adata.obs.cell_type.isin([\"JC1\", \"JC2\", \"JC3\", \"ILC3\"])]\n",
    "# Create new UMAP\n",
    "sc.tl.umap(adataSubset, min_dist=0.5, spread=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "066fa479-31eb-4667-861e-7d416f4cb959",
   "metadata": {},
   "source": [
    "### Embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7ab6418-b8f0-48f7-8aca-7d183b3b9652",
   "metadata": {},
   "source": [
    "#### sample breakdown, subclusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ab16cb-0b64-4955-9c1b-5f7a77d17456",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.pl.sc_umap(\n",
    "    adataSubset,\n",
    "    [\"sample\"],\n",
    "    cmaps=[\"tab10\"],\n",
    "    legend_loc=[ \"bottom\"],\n",
    "    latent_rep=\"X_umap\",\n",
    "    ncols=1,\n",
    "    legend_size=14,\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=14,\n",
    "    title_size=14,\n",
    "    save_path=\"analysis/JC_subset_sample_source.pdf\", \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b448058-56f8-4886-84ca-947182e2d884",
   "metadata": {},
   "source": [
    "#### Sort breakdown, subclusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbc51fa1-dd2e-427d-b6f4-fbf8bea200d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.pl.sc_umap(\n",
    "    adataSubset,\n",
    "    [\"sort\"],\n",
    "    cmaps=[\"tab10\"],\n",
    "    legend_loc=[ \"bottom\"],\n",
    "    latent_rep=\"X_umap\",\n",
    "    ncols=1,\n",
    "    legend_size=14,\n",
    "    dims=(6,6),\n",
    "    legend_fontsize=14,\n",
    "    title_size=14,\n",
    "    save_path=\"analysis/JC_subset_sort.pdf\", \n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad4b795a-ad6d-42fb-a106-5448fe6e6233",
   "metadata": {},
   "source": [
    "#### Feature plots, subclusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0039fbf8-603a-4f26-99c2-70b68e2af8dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna.pl.sc_umap(\n",
    "    adataSubset,\n",
    "    [\"cell_type\", \"Aire\", \"Rorc\", \"Itgav\", \"Itgb8\", \"Ccr7\", \"Cd80\", \"Cd86\", \"H2-Aa\", \"H2-Ab1\", \"Klf4\", \"Myc\", \"Ccr6\", \"Cd40\", \"Cxcr6\", \"Il7r\", \"Thy1\", \"Cd44\", \"Il2ra\"],\n",
    "    legend_loc=[\"None\"],\n",
    "    cmaps=[\"tab10\"],\n",
    "    gene_data=\"scVI_normalized\",\n",
    "    cmap_label=\"Normalized expression\",\n",
    "    title_size=21,\n",
    "    legend_fontsize=21,\n",
    "    dims=(6,6),\n",
    "    s=6,\n",
    "    save_path=\"analysis/WT_merged_feature_plots/littman_paper_JC_clusters.png\",\n",
    "    dpi=300,\n",
    "    thresholds=thresholds, # Use same thresholds as for all cluster data so plots are comparable\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a6e5881-6cd6-48e4-b56f-2c62b1549770",
   "metadata": {},
   "source": [
    "### Dotplots"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5a854f7-3031-4ff0-8efe-d2dbd04b4743",
   "metadata": {},
   "source": [
    "#### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f06e57-ccec-4c43-8808-47fae0c1bd81",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi=150, dpi_save=300, frameon=True)\n",
    "sc.settings.figdir = \"analysis/\"\n",
    "markers=[\"Rorc\", \"Itgav\", \"Itgb8\", \"Ccr7\", \"Cd80\", \"Cd86\", \"Aire\", \"H2-Aa\", \"H2-Ab1\",]\n",
    "sc.pl.dotplot(\n",
    "    adataSubset, \n",
    "    markers, \n",
    "    groupby='cell_type', \n",
    "    dendrogram=False, \n",
    "    use_raw=False, # Use log normalized, scaled data\n",
    "    swap_axes=True, \n",
    "    standard_scale=\"var\", # Standardize each gene to range from 0-1\n",
    "    categories_order=[\"JC1\", \"JC2\", \"JC3\", \"ILC3\"],\n",
    "    save=\"scaled_data_JC_subset_features.pdf\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef46a432-9ed2-4aff-b833-45e13546de36",
   "metadata": {},
   "source": [
    "#### TC1-4 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4160073-6461-4503-aa87-2391a8159ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi=150, dpi_save=300, frameon=True)\n",
    "plt.rcdefaults()\n",
    "sc.settings.figdir = \"analysis/\"\n",
    "markersDict={\"TC1\": [\"Nrp2\", \"Bach2\", \"Ncam1\", \"Nrxn1\", \"Cd86\", \"Cd80\", \"Nrg1\", \"Kif21a\", \"Tnfrsf11b\"], #TC1\n",
    "         \"TC2\": [\"Pde10a\", \"Cfp\", \"Hk2\", \"Nrp1\", \"Wdfy4\", \"Ltb\", \"Pigr\"], #TC2\n",
    "         \"TC3\": [\"Cd63\", \"Epsti1\", \"H2-M2\", \"Nlrc5\", \"Dnase1l3\", \"Ahcyl2\", \"Ldb2\"], #TC3\n",
    "         \"TC4\": [\"Itgav\", \"Il15ra\", \"Il2ra\", \"Marcks\", \"Cd274\", \"Ccl5\", \"Cd36\", \"Itgb8\", \"Ccl22\"], #TC4\n",
    "            }\n",
    "sc.pl.dotplot(\n",
    "    adataSubset, \n",
    "    markersDict, \n",
    "    groupby='cell_type', \n",
    "    dendrogram=False, \n",
    "    use_raw=False, # Use log normalized, scaled data\n",
    "    standard_scale=\"var\", # Standardize each gene to range from 0-1\n",
    "    categories_order=[\"JC1\", \"JC2\", \"JC3\", \"ILC3\"],\n",
    "    save=\"scaled_data_JC_subset_TC_markers.pdf\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09ac9264-1559-4579-a112-9389e4c7c1bb",
   "metadata": {},
   "source": [
    "### DE genes, subclusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a909d65d-620c-49fe-8383-0a5625264cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = scvi.model.SCVI.load(\n",
    "    \"processed_data/scVI_models/WT_LN_merged\",\n",
    "    adata\n",
    ")\n",
    "# Get DE markers for cell type clusters using scVI differential expression method\n",
    "# default thresholds for marker genes: lfc_mean > 0, bayes_factor > 3, non_zeros_proportion > 0.1\n",
    "markers = rna.tl.cluster_de(\n",
    "    adata[adata.obs.cell_type.isin([\"JC1\", \"JC2\", \"JC3\", \"ILC3\"])], # Use only JC + ILC3 subclusters for DE\n",
    "    model,\n",
    "    cluster_key=\"cell_type\",\n",
    "    save_path=\"analysis/DE_genes/WT_LN_merged/cell_type_subset_markers/\"\n",
    ")\n",
    "# Get top 15 DE genes for each cluster\n",
    "JC1_de = markers[\"JC1\"].head(15).index\n",
    "JC2_de = markers[\"JC2\"].head(15).index\n",
    "JC3_de = markers[\"JC3\"].head(15).index\n",
    "ILC3_de = markers[\"ILC3\"].head(15).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "938851e5-2ea1-488b-9a99-b0bab47663da",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = list(JC1_de)\n",
    "genes += [i for i in list(JC2_de) if i not in genes]\n",
    "genes += [i for i in list(JC3_de) if i not in genes]\n",
    "genes += [i for i in list(ILC3_de) if i not in genes]\n",
    "adataSubset.obs.cell_type.cat.reorder_categories([\"JC1\", \"JC2\", \"JC3\", \"ILC3\"], inplace=True)\n",
    "# Get normalized expression for DE genes\n",
    "expression = rna.tl.get_expression_matrix(adataSubset, \"scVI_normalized\")\n",
    "expression = expression.loc[:, genes]\n",
    "expression[\"cell_type\"] = adataSubset.obs.cell_type\n",
    "# Pseudobulk average expression of DE genes by cell type\n",
    "expression = expression.groupby(\"cell_type\").mean()\n",
    "# Clustered heatmap of data\n",
    "ax = sns.clustermap(\n",
    "    expression,\n",
    "    cmap=\"viridis\",\n",
    "    z_score=1, # Z-score each gene\n",
    "    row_cluster=False,\n",
    "    col_cluster=False,\n",
    "    cbar_pos=(1.001,0.18,0.03,0.64),\n",
    "    figsize=(23,6)\n",
    ")\n",
    "cbar = ax.ax_heatmap.collections[0].colorbar\n",
    "cbar.ax.tick_params(labelsize=20)\n",
    "ax.figure.axes[-1].set_ylabel(\"Z-score\", size=21)\n",
    "ax.ax_heatmap.tick_params(axis=\"both\", which=\"both\", labelsize=21)\n",
    "ax.ax_heatmap.yaxis.tick_left()\n",
    "ax.ax_heatmap.set_ylabel(\"\")\n",
    "plt.savefig(\"analysis/JC_DE_genes_pseudobulk.pdf\", dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5ff1bfc-9d3e-44b2-8933-7f88e8181d9b",
   "metadata": {},
   "source": [
    "### Cosine similarity, subclusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3849a221-9df2-45fc-ba38-918693c58ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can only run after the all clusters similarity cells, uses the same data but subset on clusters of interest\n",
    "dataDict = {}\n",
    "# Immgen samples of interest\n",
    "keys = [\n",
    "    'DC_8-_Th_imm_score',\n",
    "    'DC_8+_Th_imm_score',\n",
    "    'DC_4+_Sp_imm_score',\n",
    "    'DC_8+_Sp_imm_score',\n",
    "    'DC_8-4-11b-_Sp_imm_score',\n",
    "    'DC_8-4-11b+_Sp_imm_score',\n",
    "    'DC_4+_SLN_imm_score',\n",
    "    'DC_8+_SLN_imm_score',\n",
    "    'DC_8-4-11b-_SLN_imm_score',\n",
    "    'DC_8-4-11b+_SLN_imm_score',\n",
    "    'DC_4+_MLN_imm_score',\n",
    "    'DC_8+_MLN_imm_score',\n",
    "    'DC_8-4-11b-_MLN_imm_score',\n",
    "    'DC_8-4-11b+_MLN_imm_score',\n",
    "    'DC_103-11b+_Lv_imm_score',\n",
    "    'DC_103+11b-_Lv_imm_score',\n",
    "    'DC_103+11b-_LuLN_imm_score',\n",
    "    'DC_103-11b+_LuLN_imm_score',\n",
    "    'DC_103-11b+24+_Lu_imm_score',\n",
    "    'DC_103+11b-_Lu_imm_score',\n",
    "    'DC_103-11b+F4_imm_score',\n",
    "    'DC_103+11b-_SI_imm_score',\n",
    "    'DC_103+11b+_SI_imm_score',\n",
    "    'DC_pDC_8-_Sp_imm_score',\n",
    "    'DC_pDC_8+_Sp_imm_score',\n",
    "    'DC_pDC_8+_SLN_imm_score',\n",
    "    'DC_pDC_8+_MLN_imm_score',\n",
    "    'DC_IIhilang-103-11blo_SLN_imm_score',\n",
    "    'DC_IIhilang-103-11b+_SLN_imm_score',\n",
    "    'DC_IIhilang+103+11blo_SLN_imm_score',\n",
    "    'DC_IIhilang+103-11b+_SLN_imm_score',\n",
    "    'DC_LC_Sk_imm_score',\n",
    "    'Ep_MEChi_Th_imm_score',\n",
    "    'Fi_MTS15+_Th_imm_score',\n",
    "    'Fi_Sk_imm_score',\n",
    "    'FRC_MLN_imm_score',\n",
    "    'FRC_SLN_imm_score',\n",
    "    'LEC_MLN_imm_score',\n",
    "    'LEC_SLN_imm_score',\n",
    "    'BEC_MLN_imm_score',\n",
    "    'BEC_SLN_imm_score',\n",
    "    'St_31-38-44-_SLN_imm_score',\n",
    "    'proB_CLP_BM_imm_score',\n",
    "    'proB_CLP_FL_imm_score',\n",
    "    'proB_FrA_BM_imm_score',\n",
    "    'proB_FrA_FL_imm_score',\n",
    "    'proB_FrBC_BM_imm_score',\n",
    "    'proB_FrBC_FL_imm_score',\n",
    "    'preB_FrC_BM_imm_score',\n",
    "    'preB_FrD_BM_imm_score',\n",
    "    'preB_FrD_FL_imm_score',\n",
    "    'B_FrE_BM_imm_score',\n",
    "    'B_FrE_FL_imm_score',\n",
    "    'B_T1_Sp_imm_score',\n",
    "    'B_T2_Sp_imm_score',\n",
    "    'B_T3_Sp_imm_score',\n",
    "    'B_Fo_Sp_imm_score',\n",
    "    'B_Fo_MLN_imm_score',\n",
    "    'B_Fo_LN_imm_score',\n",
    "    'B_Fo_PC_imm_score',\n",
    "    'B_GC_Sp_imm_score',\n",
    "    'B_MZ_Sp_imm_score',\n",
    "    'B_FrF_BM_imm_score',\n",
    "    'B1b_PC_imm_score',\n",
    "    'B1a_PC_imm_score',\n",
    "    'B1a_Sp_imm_score',\n",
    "    'GN_BM_imm_score',\n",
    "    'GN_Bl_imm_score',\n",
    "    'MF_BM_imm_score',\n",
    "    'MF_RP_Sp_imm_score',\n",
    "    'MF_Lu_imm_score',\n",
    "    'MF_103-11b+24-_Lu_imm_score',\n",
    "    'MF_II+480lo_PC_imm_score',\n",
    "    'MF_103-11b+_SI_imm_score',\n",
    "    'MF_11cloSer_SI_imm_score',\n",
    "    'MF_II-480hi_PC_imm_score',\n",
    "    'MF_Microglia_CNS_imm_score',\n",
    "    'Mo_6C+II-_BM_imm_score',\n",
    "    'Mo_6C-II-_BM_imm_score',\n",
    "    'Mo_6C+II-_Bl_imm_score',\n",
    "    'Mo_6C+II+_Bl_imm_score',\n",
    "    'Mo_6C-II-_Bl_imm_score',\n",
    "    'Mo_6C-II+_Bl_imm_score',\n",
    "    'Mo_6C-IIint_Bl_imm_score',\n",
    "    'Mo_6C+II-_LN_imm_score',\n",
    "    'NK_Sp_imm_score',\n",
    "    'NK_49CI-_Sp_imm_score',\n",
    "    'NK_49CI+_Sp_imm_score',\n",
    "    'NK_49H-_Sp_imm_score',\n",
    "    'NK_49H+_Sp_imm_score',\n",
    "    'NK_b2m-_Sp_imm_score',\n",
    "    'NK_DAP10-_Sp_imm_score',\n",
    "    'NK_DAP12-_Sp_imm_score',\n",
    "    'NK_CD127-_Sp_imm_score',\n",
    "    'NK_CD49b+_Lv_imm_score',\n",
    "    'NK_CD127-_SI_imm_score',\n",
    "    'ILC1_CD127+_Sp_imm_score',\n",
    "    'ILC1_CD49b-_Lv_imm_score',\n",
    "    'ILC1_CD127+_SI_imm_score',\n",
    "    'ILC2_SI_imm_score',\n",
    "    'ILC3_NKp46+_Rorgthi_SI_imm_score',\n",
    "    'ILC3_NKp46-_4-_SI_imm_score',\n",
    "    'ILC3_NKp46-_4+_SI_imm_score',\n",
    "    'preT_ETP_Th_imm_score',\n",
    "    'preT_ETP-2A_Th_imm_score',\n",
    "    'preT_DN2_Th_imm_score',\n",
    "    'preT_DN2A_Th_imm_score',\n",
    "    'preT_DN2B_Th_imm_score',\n",
    "    'preT_DN2-3_Th_imm_score',\n",
    "    'preT_DN3A_Th_imm_score',\n",
    "    'preT_DN3B_Th_imm_score',\n",
    "    'preT_DN3-4_Th_imm_score',\n",
    "    'T_DN4_Th_imm_score',\n",
    "    'T_ISP_Th_imm_score',\n",
    "    'T_DP_Th_imm_score',\n",
    "    'T_DPbl_Th_imm_score',\n",
    "    'T_DPsm_Th_imm_score',\n",
    "    'T_DP69+_Th_imm_score',\n",
    "    'T_4+8int_Th_imm_score',\n",
    "    'T_4SP69+_Th_imm_score',\n",
    "    'T_4SP24int_Th_imm_score',\n",
    "    'T_4SP24-_Th_imm_score',\n",
    "    'T_4int8+_Th_imm_score',\n",
    "    'T_8SP69+_Th_imm_score',\n",
    "    'T_8SP24int_Th_imm_score',\n",
    "    'T_8SP24-_Th_imm_score',\n",
    "    'T_4Nve_Sp_imm_score',\n",
    "    'T_4Mem_Sp_imm_score',\n",
    "    'T_4Mem44h62l_Sp_imm_score',\n",
    "    'T_4Nve_LN_imm_score',\n",
    "    'T_4Mem_LN_imm_score',\n",
    "    'T_4Mem44h62l_LN_imm_score',\n",
    "    'T_4Nve_PP_imm_score',\n",
    "    'T_4Nve_MLN_imm_score',\n",
    "    'T_4_LN_BDC_imm_score',\n",
    "    'T_4_PLN_BDC_imm_score',\n",
    "    'T_4_Pa_BDC_imm_score',\n",
    "    'T_4FP3-_Sp_imm_score',\n",
    "    'T_4FP3+25+_Sp_imm_score',\n",
    "    'T_4FP3+25+_AA_imm_score',\n",
    "    'T_4FP3+25+_LN_imm_score',\n",
    "    'T_8Nve_Sp_imm_score',\n",
    "    'T_8Mem_Sp_imm_score',\n",
    "    'T_8Nve_LN_imm_score',\n",
    "    'T_8Mem_LN_imm_score',\n",
    "    'T_8Nve_PP_imm_score',\n",
    "    'T_8Nve_MLN_imm_score',\n",
    "    'NKT_44-NK1_1-_Th_imm_score',\n",
    "    'NKT_44+NK1_1-_Th_imm_score',\n",
    "    'NKT_44+NK1_1+_Th_imm_score',\n",
    "    'NKT_4+_Sp_imm_score',\n",
    "    'NKT_4-_Sp_imm_score',\n",
    "    'NKT_4+_Lv_imm_score',\n",
    "    'NKT_4-_Lv_imm_score',\n",
    "    'Tgd_Th_imm_score',\n",
    "    'Tgd_vg1+vd6-24ahi_Th_imm_score',\n",
    "    'Tgd_vg1+vd6+24ahi_Th_imm_score',\n",
    "    'Tgd_vg2+24ahi_Th_imm_score',\n",
    "    'Tgd_vg2+24ahi_e17_Th_imm_score',\n",
    "    'Tgd_vg3+24ahi_e17_Th_imm_score',\n",
    "    'Tgd_vg5+24ahi_Th_imm_score',\n",
    "    'Tgd_vg1+vd6-24alo_Th_imm_score',\n",
    "    'Tgd_vg1+vd6+24alo_Th_imm_score',\n",
    "    'Tgd_vg2+24alo_Th_imm_score',\n",
    "    'Tgd_vg3+24alo_e17_Th_imm_score',\n",
    "    'Tgd_Sp_imm_score',\n",
    "    'Tgd_vg2-_Sp_imm_score',\n",
    "    'Tgd_vg2-_act_Sp_imm_score',\n",
    "    'Tgd_vg2+_Sp_imm_score',\n",
    "    'Tgd_vg2+_act_Sp_imm_score',\n",
    "    'Tgd_vg2-_Sp_TCRbko_imm_score',\n",
    "    'Tgd_vg2+_Sp_TCRbko_imm_score',\n",
    "    'Tgd_vg5-_IEL_imm_score',\n",
    "    'Tgd_vg5+_IEL_imm_score',\n",
    "    'Tgd_vg5-_act_IEL_imm_score',\n",
    "    'Tgd_vg5+_act_IEL_imm_score',\n",
    "    ]\n",
    "for key in keys:\n",
    "    dataDict[key.rsplit(\"_\", 2)[0]] = adataCopy.obs[key]\n",
    "dataDict[\"cell_type\"] = adataCopy.obs.cell_type\n",
    "data = pd.DataFrame(dataDict)\n",
    "cats = [\"JC1\", \"JC2\", \"JC3\", \"mDC1-1\", \"mDC1-2\", \"mDC1-3\", \"mDC2-1\", \"mDC2-2\", \"mDC2-3\", \"mDC2-4\", \"mDC2-5\", \"cDC1\", \"cDC2\", \"pDC\", \"MF\", \"Mo\", \"BEC\", \"LEC\", \"Fibro1\",\n",
    "        \"Fibro2\", \"Fibro3\", \"ILC\", \"ILC3\", \"CD4 T\", \"CD8 T\", \"B\", \"Plasma cell\"]\n",
    "data.cell_type.cat.reorder_categories(cats, inplace=True)\n",
    "# Get average similarity scores for each cell type\n",
    "data = data.groupby(\"cell_type\").mean()\n",
    "# Subset on clusters of interest\n",
    "data = data[data.index.isin([\"JC1\", \"JC2\", \"JC3\", \"ILC3\"])]\n",
    "# Scale scores for each immgen sample to range from 0-1\n",
    "data = data.subtract(data.min(axis=0), axis=1)\n",
    "data = data.div(data.max(axis=0), axis=1)\n",
    "\n",
    "# Clustered heatmap of scores\n",
    "ax = sns.clustermap(\n",
    "    data,\n",
    "    cmap=\"coolwarm\",\n",
    "    row_cluster=False,\n",
    "    col_cluster=True,\n",
    "    figsize=(65,8),\n",
    "    cbar_pos=None,\n",
    ")\n",
    "ax.ax_heatmap.tick_params(axis=\"both\", which=\"both\", labelsize=21)\n",
    "ax.ax_heatmap.set_yticklabels([\"JC1\", \"JC2\", \"JC3\", \"ILC3\"], rotation=0, va=\"center\")\n",
    "ax.ax_heatmap.set_ylabel(\"\")\n",
    "ax.ax_heatmap.yaxis.tick_left()\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"analysis/JC_cluster_similarity_heatmap_clustered.png\", dpi=300)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
